File: android/src/main/java/in/cashify/otaupdate/OtaBundleManager.kt
package `in`.cashify.otaupdate

import android.content.Context
import android.util.Log
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.io.File
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArraySet

/**
 * Version selection, download and cleanup for OTA bundles.
 *
 * The LAUNCH path ([getLauncherBundleFilePathLocal]) is synchronous and
 * local-only — no Remote Config, no network — so it can run inside the host's
 * lazy `reactHost` block without blocking the first frame. Safe-mode kill
 * switches (global and per-module) from Remote Config are applied by the
 * background check and persisted via [OtaPreferences]; the launch path only
 * reads those persisted flags.
 *
 * The launcher module's shipped bundle is the host's stock flat asset
 * (`assets://index.android.bundle`), so its asset baseline version comes from
 * `moduleVersion` in app.json. Returning null hands loading back to the host's
 * default asset loader.
 */
object OtaBundleManager {

    // One lock per module so concurrent downloadBundleIfNeeded calls for the
    // SAME module serialize instead of downloading the same bundle twice.
    private val downloadMutex = ConcurrentHashMap<String, Mutex>()

    // Live progress listeners per module; an in-flight download fans out to
    // whoever is registered at the time of each network read.
    private val progressListeners =
        ConcurrentHashMap<String, CopyOnWriteArraySet<BundleDownloadProgressListener>>()

    /** Version of the disk bundle the launcher actually booted with, if any. */
    @Volatile
    var launcherLoadedBundleVersion: String? = null
        private set

    /**
     * LAUNCH PATH — decides which bundle the ReactHost boots. Returns an absolute
     * file path of a valid downloaded bundle strictly newer than the shipped one,
     * or null (-> host loads the default APK asset / Metro in debug). Never throws.
     */
    fun getLauncherBundleFilePathLocal(context: Context): String? {
        val tag = "OtaBundleManager::getLauncherBundleFilePathLocal"
        return try {
            if (HostAppInfo.isDebuggable(context)) {
                Log.d("CashifyOTA", "$tag::debuggable build, using default bundle source")
                return null
            }
            if (OtaPreferences.isSafeModeEnabled(context)) {
                Log.d("CashifyOTA", "$tag::safe mode enabled, using asset bundle")
                return null
            }
            val module = OtaModuleManager.launcherModule()
            if (module == null) {
                Log.d("CashifyOTA", "$tag::no launcher module configured, using asset bundle")
                return null
            }
            if (OtaPreferences.isModuleSafeModeEnabled(context, module.configKey)) {
                Log.d("CashifyOTA", "$tag::module safe mode enabled for ${module.moduleName}, using asset bundle")
                return null
            }
            val assetVersion = getJsBundleAssetVersion(module)
            val diskVersion = getJsBundleDiskVersion(context, module)
            if (diskVersion == null) {
                Log.d("CashifyOTA", "$tag::no valid disk bundle, using asset bundle ($assetVersion)")
                OtaModuleManager.setBundleVersion(module.moduleName, assetVersion ?: "")
                return null
            }
            // Asset wins ties: disk must be STRICTLY newer than the shipped bundle.
            if (assetVersion != null && listOf(diskVersion, assetVersion).semanticMax() == assetVersion) {
                Log.d("CashifyOTA", "$tag::disk $diskVersion <= asset $assetVersion, using asset bundle")
                OtaModuleManager.setBundleVersion(module.moduleName, assetVersion)
                return null
            }
            val bundleFile = buildJsBundleFilePath(context, module, diskVersion)
            launcherLoadedBundleVersion = diskVersion
            OtaModuleManager.setBundleVersion(module.moduleName, diskVersion)
            Log.d("CashifyOTA", "$tag::loading disk bundle $diskVersion: ${bundleFile.absolutePath}")
            bundleFile.absolutePath
        } catch (t: Throwable) {
            Log.e("CashifyOTA", "$tag::failed, falling back to asset bundle: ${t.message}", t)
            null
        }
    }

    /**
     * Resolves any module's bundle for the ScriptResolver bridge: valid disk
     * bundle first, launcher's flat asset as fallback. Throws when nothing exists
     * (non-launcher module with no downloaded bundle, or module safe-mode on a
     * non-launcher module).
     */
    fun getJsBundleUriForModule(context: Context, moduleName: String): String {
        val tag = "OtaBundleManager::getJsBundleUriForModule::$moduleName"
        val module = OtaModuleManager.getModuleByName(moduleName)
            ?: throw IllegalArgumentException("$tag::module not found")

        val safeMode = OtaPreferences.isSafeModeEnabled(context) ||
            OtaPreferences.isModuleSafeModeEnabled(context, module.configKey)
        if (!safeMode) {
            val diskVersion = getJsBundleDiskVersion(context, module)
            val assetVersion = getJsBundleAssetVersion(module)
            if (diskVersion != null &&
                (assetVersion == null || listOf(diskVersion, assetVersion).semanticMax() == diskVersion) &&
                diskVersion != assetVersion
            ) {
                val bundleFile = buildJsBundleFilePath(context, module, diskVersion)
                OtaModuleManager.setBundleVersion(module.moduleName, diskVersion)
                Log.d("CashifyOTA", "$tag::disk bundle $diskVersion")
                return bundleFile.absolutePath
            }
        }
        if (module.launcher) {
            OtaModuleManager.setBundleVersion(module.moduleName, module.moduleVersion)
            Log.d("CashifyOTA", "$tag::asset bundle ${module.moduleVersion}")
            return "assets://${getBundleName()}"
        }
        throw IllegalStateException("$tag::no bundle available for non-launcher module")
    }

    suspend fun downloadBundleIfNeeded(
        context: Context,
        module: OtaModule,
        onProgress: BundleDownloadProgressListener? = null
    ) {
        val tag = "OtaBundleManager::downloadBundleIfNeeded::${module.moduleName}"
        Log.d("CashifyOTA", tag)
        // Registered BEFORE waiting on the mutex so a caller that joins while
        // another caller's download is in flight still receives that download's
        // progress.
        val listeners = progressListeners.getOrPut(module.moduleName) { CopyOnWriteArraySet() }
        onProgress?.let { listeners.add(it) }
        val fanOutProgress: BundleDownloadProgressListener = { bytesRead, totalBytes, done ->
            listeners.forEach { listener ->
                try {
                    listener(bytesRead, totalBytes, done)
                } catch (e: Exception) {
                    Log.w("CashifyOTA", "$tag::progress listener failed", e)
                }
            }
        }
        try {
            val mutex = downloadMutex.getOrPut(module.moduleName) { Mutex() }
            if (mutex.isLocked) {
                Log.d("CashifyOTA", "$tag::download already in flight, waiting for it to finish")
            }
            mutex.withLock {
                try {
                    val remoteVersion = OtaRemoteConfig.getBundleLatestVersion(context, module)
                    Log.d("CashifyOTA", "$tag::remoteVersion: $remoteVersion")
                    if (remoteVersion.isEmpty()) {
                        Log.d("CashifyOTA", "$tag::remoteVersion is empty, skip download")
                        return
                    }
                    if (!remoteVersion.isSemanticVersion()) {
                        Log.w("CashifyOTA", "$tag::remoteVersion '$remoteVersion' is not a valid version, skip download")
                        return
                    }

                    val assetVersion = getJsBundleAssetVersion(module)
                    Log.d("CashifyOTA", "$tag::assetVersion: $assetVersion")
                    if (assetVersion != null &&
                        listOf(assetVersion, remoteVersion).semanticMax() == assetVersion
                    ) {
                        Log.d("CashifyOTA", "$tag::assetVersion >= remoteVersion, skip download")
                        return
                    }

                    val diskVersion = getJsBundleDiskVersion(context, module)
                    if (diskVersion != null &&
                        listOf(diskVersion, remoteVersion).semanticMax() == diskVersion
                    ) {
                        Log.d("CashifyOTA", "$tag::diskVersion $diskVersion >= remoteVersion, skip download")
                        return
                    }

                    val bundleUrl = OtaRemoteConfig.getBundleUrl(context)
                    if (bundleUrl.isEmpty()) {
                        Log.d("CashifyOTA", "$tag::rn_bundle_url is empty, OTA disabled, skip download")
                        return
                    }

                    Log.d("CashifyOTA", "$tag::downloading remote bundle $remoteVersion")
                    downloadRemoteBundle(context, tag, module, bundleUrl, remoteVersion, fanOutProgress)
                } catch (e: Exception) {
                    Log.e("CashifyOTA", "$tag::error: ${e.message}", e)
                }
            }
        } finally {
            // Runs on every exit path so listeners never leak.
            onProgress?.let { listeners.remove(it) }
        }
    }

    suspend fun cleanupStaleBundles(context: Context, module: OtaModule) {
        val tag = "OtaBundleManager::cleanupStaleBundles::${module.moduleName}"
        val moduleDir = File(getFilesDir(context), module.modulePath)
        if (!moduleDir.exists()) return
        val versionDirs = moduleDir.listFiles()?.filter { it.isDirectory } ?: emptyList()
        if (versionDirs.isEmpty()) return

        val assetVersion = getJsBundleAssetVersion(module)
        val remoteVersion = try {
            OtaRemoteConfig.getBundleLatestVersion(context, module)
        } catch (e: Exception) {
            ""
        }
        Log.d("CashifyOTA", "$tag::assetVersion: $assetVersion, remoteVersion: $remoteVersion")

        val invalidVersions = mutableListOf<String>()
        for (versionDir in versionDirs) {
            // empty directory
            if (versionDir.listFiles()?.isEmpty() == true) {
                Log.d("CashifyOTA", "$tag::empty version directory: ${versionDir.name}")
                invalidVersions.add(versionDir.name)
                versionDir.deleteRecursively()
                continue
            }
            // rollback: remote was lowered below this version
            if (remoteVersion.isNotEmpty() && remoteVersion != versionDir.name &&
                listOf(remoteVersion, versionDir.name).semanticMin() == remoteVersion
            ) {
                Log.d("CashifyOTA", "$tag::rollback detected, deleting: ${versionDir.name}")
                invalidVersions.add(versionDir.name)
                versionDir.deleteRecursively()
                continue
            }
            // stale: not newer than the shipped bundle
            if (assetVersion != null &&
                listOf(assetVersion, versionDir.name).semanticMax() == assetVersion
            ) {
                Log.d("CashifyOTA", "$tag::stale version detected, deleting: ${versionDir.name}")
                invalidVersions.add(versionDir.name)
                versionDir.deleteRecursively()
                continue
            }
            val bundleFile = File(versionDir, getBundleName())
            if (!bundleFile.exists() || !BundleDownloader.hasMarker(bundleFile)) {
                Log.d("CashifyOTA", "$tag::missing/corrupt bundle, deleting: ${versionDir.name}")
                invalidVersions.add(versionDir.name)
                versionDir.deleteRecursively()
                continue
            }
        }

        val validVersionDirs = versionDirs.filter { !invalidVersions.contains(it.name) }
        if (validVersionDirs.size <= 1) return
        // Keep only the newest valid version.
        val latestVersion = validVersionDirs.map { it.name }.semanticMax()
        for (versionDir in validVersionDirs) {
            if (versionDir.name != latestVersion) {
                Log.d("CashifyOTA", "$tag::deleting old version directory: ${versionDir.name}")
                versionDir.deleteRecursively()
            }
        }
    }

    fun cleanupTemporaryBundles(context: Context) {
        val tag = "OtaBundleManager::cleanupTemporaryBundles"
        val tempFiles = getTemporaryDir(context).listFiles()?.filter { it.name.endsWith(".tmp") }
            ?: emptyList()
        for (tempFile in tempFiles) {
            try {
                tempFile.delete()
                Log.d("CashifyOTA", "$tag::deleted temporary file: ${tempFile.absolutePath}")
            } catch (e: Exception) {
                Log.e("CashifyOTA", "$tag::failed to delete ${tempFile.absolutePath}: ${e.message}")
            }
        }
    }

    fun cleanupModuleBundles(context: Context, module: OtaModule) {
        val tag = "OtaBundleManager::cleanupModuleBundles::${module.moduleName}"
        val moduleDir = File(getFilesDir(context), module.modulePath)
        if (moduleDir.exists()) {
            val result = moduleDir.deleteRecursively()
            Log.d("CashifyOTA", "$tag::deleted module directory: ${moduleDir.absolutePath} result: $result")
        }
    }

    private fun downloadRemoteBundle(
        context: Context,
        tag: String,
        module: OtaModule,
        bundleUrl: String,
        remoteVersion: String,
        onProgress: BundleDownloadProgressListener? = null
    ) {
        val destinationFile = buildJsBundleFilePath(context, module, remoteVersion)
        val tempFile = File(getTemporaryDir(context), UUID.randomUUID().toString() + ".tmp")
        val remoteUrl = buildJsBundleRemoteUrl(bundleUrl, module, remoteVersion)

        val downloadFilePath = BundleDownloader.downloadFileSync(remoteUrl, tempFile, onProgress)
        if (downloadFilePath == null) {
            if (tempFile.exists()) tempFile.delete()
            throw Exception("$tag::failed to download bundle from $remoteUrl")
        }

        if (!BundleDownloader.hasMarker(tempFile)) {
            Log.d("CashifyOTA", "$tag::downloaded bundle has no marker, deleting")
            tempFile.delete()
            return
        }

        try {
            destinationFile.parentFile?.let { if (!it.exists()) it.mkdirs() }
            // cacheDir and filesDir share a filesystem, so this rename is atomic.
            val renamed = tempFile.renameTo(destinationFile)
            if (!renamed) {
                Log.e("CashifyOTA", "$tag::renameTo failed, cleaning up")
                tempFile.delete()
                if (destinationFile.exists()) destinationFile.delete()
                return
            }
            Log.d("CashifyOTA", "$tag::bundle installed at: ${destinationFile.absolutePath}")
        } catch (e: Exception) {
            Log.e("CashifyOTA", "$tag::error moving file: ${e.message}")
            tempFile.delete()
            if (destinationFile.exists()) destinationFile.delete()
        }
    }

    private fun buildJsBundleRemoteUrl(bundleUrl: String, module: OtaModule, version: String): String {
        return "${bundleUrl.trimEnd('/')}/${module.modulePath}/$version/${getBundleName()}.zip".also {
            Log.d("CashifyOTA", "OtaBundleManager::buildJsBundleRemoteUrl::url: $it")
        }
    }

    /**
     * The version of the bundle shipped inside this binary. The launcher's flat
     * asset carries `moduleVersion` from app.json; non-launcher modules ship no
     * asset at all (disk/remote only).
     */
    internal fun getJsBundleAssetVersion(module: OtaModule): String? {
        return if (module.launcher) module.moduleVersion else null
    }

    private fun getJsBundleDiskVersion(context: Context, module: OtaModule): String? {
        return try {
            val versionFolders = File(getFilesDir(context), module.modulePath).list()
            val validVersions = versionFolders?.filter { version ->
                val bundleFile = buildJsBundleFilePath(context, module, version)
                bundleFile.exists() && BundleDownloader.hasMarker(bundleFile)
            } ?: emptyList()
            validVersions.semanticMax()
        } catch (e: Exception) {
            Log.e("CashifyOTA", "OtaBundleManager::getJsBundleDiskVersion error: ${e.message}")
            null
        }
    }

    private fun buildJsBundleFilePath(context: Context, module: OtaModule, version: String): File {
        return File(getFilesDir(context), "${module.modulePath}/$version/${getBundleName()}")
    }

    private fun getFilesDir(context: Context): File = context.filesDir

    private fun getTemporaryDir(context: Context): File = context.cacheDir

    internal fun getBundleName(): String = "index.android.bundle"
}

File: ios/OtaBundleManager.swift
import Foundation

/// Version selection, download and cleanup for OTA bundles.
///
/// The LAUNCH path (`getLauncherBundleFileURLLocal`) is synchronous and
/// local-only — no Remote Config, no network — so the host's `bundleURL()`
/// never blocks the first frame. Safe-mode kill switches (global and
/// per-module) from Remote Config are applied by the background check and
/// persisted via `OtaPreferences`; the launch path only reads those flags.
///
/// The launcher module's shipped bundle is the host's stock flat
/// `main.jsbundle`, so its asset baseline version comes from `moduleVersion`
/// in app.json. Returning nil hands loading back to the default asset bundle.
///
/// Storage: Application Support/<modulePath>/<version>/main.jsbundle (excluded
/// from iCloud backup).
@objc public class OtaBundleManager: NSObject {

  /// Version of the disk bundle the launcher actually booted with, if any.
  private(set) static var loadedBundleVersion: String?

  /// OTA version when a disk bundle booted, else the host app's versionName
  /// (the shipped asset bundle).
  @objc public static func currentBundleVersion() -> String {
    loadedBundleVersion
      ?? (Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.0")
  }

  // MARK: - Launch path (synchronous, local-only, never throws)

  public static func getLauncherBundleFileURLLocal() -> URL? {
    let tag = "OtaBundleManager::getLauncherBundleFileURLLocal"
#if DEBUG
    Log.d("\(tag)::debug build, using default bundle source")
    return nil
#else
    if OtaPreferences.isSafeModeEnabled {
      Log.d("\(tag)::safe mode enabled, using asset bundle")
      return nil
    }
    guard let module = OtaModuleManager.shared.launcherModule() else {
      Log.d("\(tag)::no launcher module configured, using asset bundle")
      return nil
    }
    if OtaPreferences.isModuleSafeModeEnabled(module.configKey) {
      Log.d("\(tag)::module safe mode enabled for \(module.moduleName), using asset bundle")
      return nil
    }
    let assetVersion = getJsBundleAssetVersion(module: module)
    guard let diskVersion = getJsBundleDiskVersion(module: module) else {
      Log.d("\(tag)::no valid disk bundle, using asset bundle (\(assetVersion ?? "nil"))")
      OtaModuleManager.shared.setBundleVersion(moduleName: module.moduleName, version: assetVersion ?? "")
      return nil
    }
    // Asset wins ties: disk must be STRICTLY newer than the shipped bundle.
    if let assetVersion, [diskVersion, assetVersion].semanticMax() == assetVersion {
      Log.d("\(tag)::disk \(diskVersion) <= asset \(assetVersion), using asset bundle")
      OtaModuleManager.shared.setBundleVersion(moduleName: module.moduleName, version: assetVersion)
      return nil
    }
    guard let bundleFile = try? buildJsBundleFilePath(modulePath: module.modulePath, version: diskVersion) else {
      return nil
    }
    loadedBundleVersion = diskVersion
    OtaModuleManager.shared.setBundleVersion(moduleName: module.moduleName, version: diskVersion)
    Log.d("\(tag)::loading disk bundle \(diskVersion): \(bundleFile.path)")
    return bundleFile
#endif
  }

  /// Bridge entry for `NativeModules.CashifyOtaUpdate.getFileSystemURL` —
  /// completion form so the ObjC module shim needs no Swift/React interop types.
  @objc public static func fileSystemURL(
    forModule moduleName: String,
    completion: @escaping (NSString?, NSError?) -> Void
  ) {
    do {
      let url = try getJsBundleUriForModule(moduleName: moduleName)
      completion(url.absoluteString as NSString, nil)
    } catch {
      completion(nil, error as NSError)
    }
  }

  /// Resolves any module's bundle: valid disk bundle first, launcher's flat
  /// asset as fallback. Throws when nothing exists (non-launcher module with no
  /// downloaded bundle, or module safe-mode on a non-launcher module).
  static func getJsBundleUriForModule(moduleName: String) throws -> URL {
    let tag = "OtaBundleManager::getJsBundleUriForModule::\(moduleName)"
    guard let module = OtaModuleManager.shared.getModuleByName(moduleName) else {
      throw OtaBundleManagerError.invalidModule(reason: moduleName)
    }

    let safeMode = OtaPreferences.isSafeModeEnabled
      || OtaPreferences.isModuleSafeModeEnabled(module.configKey)
    if !safeMode,
       let diskVersion = getJsBundleDiskVersion(module: module) {
      let assetVersion = getJsBundleAssetVersion(module: module)
      if assetVersion == nil || [diskVersion, assetVersion!].semanticMax() == diskVersion,
         diskVersion != assetVersion,
         let bundleFile = try? buildJsBundleFilePath(modulePath: module.modulePath, version: diskVersion) {
        OtaModuleManager.shared.setBundleVersion(moduleName: module.moduleName, version: diskVersion)
        Log.d("\(tag)::disk bundle \(diskVersion)")
        return bundleFile
      }
    }
    if module.launcher, let assetUrl = Bundle.main.url(forResource: "main", withExtension: "jsbundle") {
      OtaModuleManager.shared.setBundleVersion(moduleName: module.moduleName, version: module.moduleVersion)
      Log.d("\(tag)::asset bundle \(module.moduleVersion)")
      return assetUrl
    }
    throw OtaBundleManagerError.invalidBundle(reason: "no bundle available for \(moduleName)")
  }

  // MARK: - Background path

  static func downloadBundleIfNeeded(module: OtaModule) async {
    let tag = "OtaBundleManager::downloadBundleIfNeeded::\(module.moduleName)"
    do {
      let remoteVersion = await OtaRemoteConfig.getBundleLatestVersion(module: module)
      Log.d("\(tag)::remoteVersion: \(remoteVersion)")
      if remoteVersion.isEmpty {
        Log.d("\(tag)::remoteVersion is empty, skip download")
        return
      }
      if !remoteVersion.isSemanticVersion {
        Log.e("\(tag)::remoteVersion '\(remoteVersion)' is not a valid version, skip download")
        return
      }

      let assetVersion = getJsBundleAssetVersion(module: module)
      if let assetVersion, [assetVersion, remoteVersion].semanticMax() == assetVersion {
        Log.d("\(tag)::assetVersion \(assetVersion) >= remoteVersion, skip download")
        return
      }

      if let diskVersion = getJsBundleDiskVersion(module: module),
         [diskVersion, remoteVersion].semanticMax() == diskVersion {
        Log.d("\(tag)::diskVersion \(diskVersion) >= remoteVersion, skip download")
        return
      }

      let bundleUrl = await OtaRemoteConfig.getBundleUrl()
      guard !bundleUrl.isEmpty else {
        Log.d("\(tag)::rn_bundle_url is empty, OTA disabled, skip download")
        return
      }

      Log.d("\(tag)::downloading remote bundle \(remoteVersion)")
      try await downloadRemoteBundle(module: module, bundleUrl: bundleUrl, remoteVersion: remoteVersion)
    } catch {
      Log.e("\(tag)::error: \(error.localizedDescription)")
    }
  }

  static func cleanupStaleBundles(module: OtaModule) async {
    let tag = "OtaBundleManager::cleanupStaleBundles::\(module.moduleName)"
    let fileManager = FileManager.default
    guard let moduleDir = try? getFilesDir(modulePath: module.modulePath),
          fileManager.fileExists(atPath: moduleDir.path) else {
      return
    }
    do {
      var versionDirs = try fileManager.contentsOfDirectory(
        at: moduleDir, includingPropertiesForKeys: nil, options: .skipsHiddenFiles)
      guard !versionDirs.isEmpty else { return }

      let assetVersion = getJsBundleAssetVersion(module: module)
      let remoteVersion = await OtaRemoteConfig.getBundleLatestVersion(module: module)
      Log.d("\(tag)::assetVersion: \(assetVersion ?? "nil"), remoteVersion: \(remoteVersion)")

      var invalidVersions: [String] = []
      for versionDir in versionDirs {
        let name = versionDir.lastPathComponent
        let contents = try? fileManager.contentsOfDirectory(
          at: versionDir, includingPropertiesForKeys: nil, options: .skipsHiddenFiles)
        // empty directory
        if let contents, contents.isEmpty {
          Log.d("\(tag)::empty version directory: \(name)")
          try? fileManager.removeItem(at: versionDir)
          invalidVersions.append(name)
          continue
        }
        // rollback: remote was lowered below this version
        if !remoteVersion.isEmpty, remoteVersion != name,
           [remoteVersion, name].semanticMin() == remoteVersion {
          Log.d("\(tag)::rollback detected, deleting: \(name)")
          try? fileManager.removeItem(at: versionDir)
          invalidVersions.append(name)
          continue
        }
        // stale: not newer than the shipped bundle
        if let assetVersion, [assetVersion, name].semanticMax() == assetVersion {
          Log.d("\(tag)::stale version detected, deleting: \(name)")
          try? fileManager.removeItem(at: versionDir)
          invalidVersions.append(name)
          continue
        }
        let bundleFile = versionDir.appendingPathComponent(getBundleName())
        if !fileManager.fileExists(atPath: bundleFile.path)
            || !BundleDownloader.hasMarker(filePath: bundleFile.path) {
          Log.d("\(tag)::missing/corrupt bundle, deleting: \(name)")
          try? fileManager.removeItem(at: versionDir)
          invalidVersions.append(name)
          continue
        }
      }

      versionDirs = versionDirs.filter { !invalidVersions.contains($0.lastPathComponent) }
      guard versionDirs.count > 1 else { return }
      // Keep only the newest valid version.
      guard let latestVersion = versionDirs.map(\.lastPathComponent).semanticMax() else { return }
      for versionDir in versionDirs where versionDir.lastPathComponent != latestVersion {
        Log.d("\(tag)::deleting old version directory: \(versionDir.lastPathComponent)")
        try? fileManager.removeItem(at: versionDir)
      }
    } catch {
      Log.e("\(tag)::error: \(error.localizedDescription)")
    }
  }

  static func cleanupTemporaryBundles() {
    let tag = "OtaBundleManager::cleanupTemporaryBundles"
    let fileManager = FileManager.default
    let tempDir = fileManager.temporaryDirectory
    guard let tempFiles = try? fileManager.contentsOfDirectory(
      at: tempDir, includingPropertiesForKeys: nil, options: .skipsHiddenFiles) else {
      return
    }
    for tempFile in tempFiles where tempFile.lastPathComponent.hasSuffix(".tmp") {
      do {
        try fileManager.removeItem(at: tempFile)
        Log.d("\(tag)::deleted temporary file: \(tempFile.path)")
      } catch {
        Log.e("\(tag)::failed to delete \(tempFile.path): \(error.localizedDescription)")
      }
    }
  }

  static func cleanupModuleBundles(module: OtaModule) {
    let tag = "OtaBundleManager::cleanupModuleBundles::\(module.moduleName)"
    let fileManager = FileManager.default
    guard let moduleDir = try? getFilesDir(modulePath: module.modulePath),
          fileManager.fileExists(atPath: moduleDir.path) else {
      return
    }
    do {
      try fileManager.removeItem(at: moduleDir)
      Log.d("\(tag)::deleted module directory: \(moduleDir.path)")
    } catch {
      Log.e("\(tag)::failed to delete \(moduleDir.path): \(error.localizedDescription)")
    }
  }

  private static func downloadRemoteBundle(
    module: OtaModule,
    bundleUrl: String,
    remoteVersion: String
  ) async throws {
    let tag = "OtaBundleManager::downloadRemoteBundle::\(module.moduleName)"
    let fileManager = FileManager.default
    let destinationFile = try buildJsBundleFilePath(modulePath: module.modulePath, version: remoteVersion)
    let tempFile = fileManager.temporaryDirectory.appendingPathComponent(UUID().uuidString + ".tmp")

    guard let remoteUrl = buildJsBundleRemoteUrl(
      bundleUrl: bundleUrl, module: module, version: remoteVersion) else {
      throw OtaBundleManagerError.downloadFailed(reason: "invalid remote URL")
    }

    guard await BundleDownloader.downloadFileSync(apiUrl: remoteUrl, bundleFile: tempFile) != nil else {
      try? fileManager.removeItem(at: tempFile)
      throw OtaBundleManagerError.downloadFailed(reason: "failed to download \(remoteUrl.absoluteString)")
    }

    guard BundleDownloader.hasMarker(filePath: tempFile.path) else {
      Log.d("\(tag)::downloaded bundle has no marker, deleting")
      try? fileManager.removeItem(at: tempFile)
      return
    }

    do {
      let destinationDir = destinationFile.deletingLastPathComponent()
      if !fileManager.fileExists(atPath: destinationDir.path) {
        try fileManager.createDirectory(at: destinationDir, withIntermediateDirectories: true)
      }
      excludeFromBackup(try getFilesDir(modulePath: module.modulePath))
      if fileManager.fileExists(atPath: destinationFile.path) {
        _ = try fileManager.replaceItemAt(destinationFile, withItemAt: tempFile)
      } else {
        try fileManager.moveItem(at: tempFile, to: destinationFile)
      }
      Log.d("\(tag)::bundle installed at: \(destinationFile.path)")
    } catch {
      Log.e("\(tag)::error installing bundle: \(error.localizedDescription)")
      try? fileManager.removeItem(at: tempFile)
      try? fileManager.removeItem(at: destinationFile)
    }
  }

  // MARK: - Paths & versions

  /// The version of the bundle shipped inside this binary. The launcher's flat
  /// asset carries `moduleVersion` from app.json; non-launcher modules ship no
  /// asset at all (disk/remote only).
  private static func getJsBundleAssetVersion(module: OtaModule) -> String? {
    module.launcher ? module.moduleVersion : nil
  }

  private static func getJsBundleDiskVersion(module: OtaModule) -> String? {
    let fileManager = FileManager.default
    guard let moduleDir = try? getFilesDir(modulePath: module.modulePath),
          let versionDirs = try? fileManager.contentsOfDirectory(
            at: moduleDir, includingPropertiesForKeys: nil, options: .skipsHiddenFiles) else {
      return nil
    }
    let validVersions = versionDirs
      .map(\.lastPathComponent)
      .filter { version in
        guard let bundleFile = try? buildJsBundleFilePath(modulePath: module.modulePath, version: version) else {
          return false
        }
        return fileManager.fileExists(atPath: bundleFile.path)
          && BundleDownloader.hasMarker(filePath: bundleFile.path)
      }
    return validVersions.semanticMax()
  }

  private static func buildJsBundleRemoteUrl(bundleUrl: String, module: OtaModule, version: String) -> URL? {
    let base = bundleUrl.hasSuffix("/") ? String(bundleUrl.dropLast()) : bundleUrl
    let url = URL(string: "\(base)/\(module.modulePath)/\(version)/\(getBundleName()).zip")
    Log.d("OtaBundleManager::buildJsBundleRemoteUrl::url: \(url?.absoluteString ?? "nil")")
    return url
  }

  private static func buildJsBundleFilePath(modulePath: String, version: String) throws -> URL {
    try getFilesDir(modulePath: modulePath)
      .appendingPathComponent(version)
      .appendingPathComponent(getBundleName())
  }

  /// Application Support/<modulePath> — NOT Documents (user-visible + backed
  /// up); backup exclusion is set when the directory is created.
  private static func getFilesDir(modulePath: String) throws -> URL {
    try FileManager.default
      .url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
      .appendingPathComponent(modulePath)
  }

  private static func excludeFromBackup(_ url: URL) {
    var mutableUrl = url
    var values = URLResourceValues()
    values.isExcludedFromBackup = true
    do {
      try mutableUrl.setResourceValues(values)
    } catch {
      Log.e("OtaBundleManager::excludeFromBackup failed for \(url.path): \(error.localizedDescription)")
    }
  }

  private static func getBundleName() -> String {
    "main.jsbundle"
  }
}

enum OtaBundleManagerError: Error {
  case invalidModule(reason: String)
  case invalidBundle(reason: String)
  case downloadFailed(reason: String)
}
